src/ips.rs

```diff
@@ -0,0 +1,445 @@
+            .geolocation
+            .as_ref()
+            .map(|geo_info| {
+                Location::new(
+                    geo_info.latitude.unwrap_or_default(), // TODO(asmie): refactor when zg-core will be updated with Location
+                    geo_info.longitude.unwrap_or_default(),
+                )
+            })
+            .unwrap_or(Location::new(0.0, 0.0));
+
+        for (node_idx, node) in nodes.iter().enumerate() {
+            let mut rating = 0.0;
+            if let Some(geo_info) = &node.geolocation {
+                let location = Location::new(
+                    geo_info.latitude.unwrap_or_default(),
+                    geo_info.longitude.unwrap_or_default(),
+                );
+                let distance = selected_location.distance_to(&location).unwrap().meters();
+                let pref_distance = self.config.pref_location_distance as f64 * 1000.0;
+
+                // Map distance to some levels of rating - now they are taken arbitrarily but
+                // they should be somehow related to the distance.
+                if self.config.pref_location_closer {
+                    match distance {
+                        _ if distance < pref_distance => rating = NORMALIZE_TO_VALUE,
+                        _ if distance < 2.0 * pref_distance => {
+                            rating = NORMALIZE_TO_VALUE * 2.0 / 3.0
+                        }
+                        _ if distance < 3.0 * pref_distance => {
+                            rating = NORMALIZE_TO_VALUE * 1.0 / 3.0
+                        }
+                        _ => rating = 0.0,
+                    }
+                } else {
+                    match distance {
+                        _ if distance < 0.5 * pref_distance => rating = 0.0,
+                        _ if distance < pref_distance => rating = NORMALIZE_TO_VALUE / 2.0,
+                        _ => rating = NORMALIZE_TO_VALUE,
+                    }
+                }
+            }
+            ratings[node_idx].2 += rating * self.config.location_weight;
+        }
+    }
+
+    fn construct_graph(&self, nodes: &[Node], agraph: &AGraph) -> Graph<IpAddr> {
+        let mut graph = Graph::new();
+
+        for i in 0..agraph.len() {
+            for j in 0..agraph[i].len() {
+                let edge = Edge::new(
+                    IpAddr::from_str(nodes[i].ip.as_str()).unwrap(),
+                    IpAddr::from_str(nodes[j].ip.as_str()).unwrap(),
+                );
+                graph.insert(edge);
+            }
+        }
+        graph
+    }
+
+    fn determine_degrees_factors(&mut self, degrees: &HashMap<IpAddr, u32>) {
+        let min = *degrees
+            .iter()
+            .min_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap_or(&0);
+        let max = *degrees
+            .iter()
+            .max_by(|a, b| a.1.cmp(b.1))
+            .map(|m| m.1)
+            .unwrap_or(&(NORMALIZE_TO_VALUE as u32));
+
```

Comment:
For here, and the other similar derivation or min/max property values, we don't want to use the normalization value as defaults: 0 and 100. They only have meaning for normalized values, not for the raw min/max to be found here and similar places.  This should always work, so panic would be better, as sign of a real bug or something else.

Reply:
Nice catch!
Done :)